StaticMoveType follow ground

saurtron opened the pull request:
### Work done

- Make StaticMoveType units follow ground right after heightmap is changed instead of polling all the time inside SlowUpdate.
- Optimize out uneeded Move() calls by checking for required change size.

### Related issues

- Fixes slow reaction to terrain change as described here: https://github.com/beyond-all-reason/RecoilEngine/pull/2182#issuecomment-2801775086

### Remarks

- May still need some work on update method
- pros: Can likely avoid some processing when https://github.com/beyond-all-reason/RecoilEngine/pull/2182#issuecomment-2839318561 is used.
- cons: Will make terrain deformation a bit more expensive.
  - will be helped by merging damage areas before notifying, rn situation is a bit bad related to that, since could get lots of updates with same or very similar rects (sometimes in same frame, but very often in consecutive frames, maybe due to projectile slowUpdate?).
  - can be helped too by reusing the quadField query among other damage listeners (like the featureHandler)
- Haven't been able to fully test, but I think this will be a net benefit in perf, as well as more correct.
  - can maybe merge update+terrain update request through updating (and renaming) current `firstUpdate` flag.
- Maintaining an initial `SlowUpdate` with `FitToGround` call, but I think it shouldn't be needed.

### Screenshots

with this change

![terraform3](https://github.com/user-attachments/assets/6450a68a-7ab7-41d3-9caf-a5b01f779bc6)

see the previous situation here: https://github.com/beyond-all-reason/RecoilEngine/pull/2182#issuecomment-2801775086

the above version (the one in this PR v1), updates on TerrainUpdated, potentially generating lots of sync events, and also somewhat lags behind where buildings can go below the floor at times.

other interpolation options (not yet in PR, **still evaluating**) with somewhat reduced sync events, also with smoother update via `moveType->Update` instead of moveType->SlowUpdate (with SlowUpdate will always lag behind).

`CGround::GetHeightReal` on update:

![det-interpolation](https://github.com/user-attachments/assets/777c500a-21a1-48f8-bb32-fe71c914992a)

`CGround::GetApproximateHeight` update, and skip `preFramePos`:

![det-nointerpolation](https://github.com/user-attachments/assets/7c74c73e-5532-4d8e-890f-05557d4e98ca)

saurtron committed:
Make StaticMoveType units follow map damage automatically, instead of needing to poll inside SlowUpdate.

saurtron committed:
Don't move if not needed.

sprunk commented on rts/Sim/MoveTypes/StaticMoveType.cpp:
```diff
@@ -32,17 +35,44 @@ CStaticMoveType::CStaticMoveType(CUnit* unit) : AMoveType(unit) {
-		owner->Move(UpVector * (CGround::GetHeightReal(owner->pos.x, owner->pos.z) - owner->pos.y), true);
+		change = CGround::GetHeightReal(owner->pos.x, owner->pos.z) - owner->pos.y;
 	}
+	if (std::abs(change) > 1e-04f)
```
there is some sort of `float3::eps` or something like that, perhaps use that for the epsilon?

saurtron replied:
done

sprunk commented on rts/Sim/MoveTypes/StaticMoveType.cpp:
```diff
@@ -32,17 +35,44 @@ CStaticMoveType::CStaticMoveType(CUnit* unit) : AMoveType(unit) {
 {
 	RECOIL_DETAILED_TRACY_ZONE;
-	// buildings and pseudo-static units can be transported
+	if (firstUpdate) {
+		FitToGround();
+		firstUpdate = false;
+	}
```
Why not just do this first update in unit created?

saurtron replied:
Not totally sure this is needed, removed it for now. if needed can add something to unit created as you recommend.

saurtron committed:
Use GetApproximateHeight for ground fit, and fit on unit->Update.

saurtron committed:
Use float3::cmp_eps() instead of hardcoded value.

saurtron commented:
Changed the fit method, to be done at `unit->Update`.

This works better with terraforming, since that happens inside `SlowUpdate`, that's run just before `Update`, this is the order of updates:

```cpp
UpdateUnitMoveTypes
SlowUpdateUnits
UpdateUnits
```

This means updating on moveType directly, even in `Update`, will always come one frame later. By doing it inside Update this can be properly synchronized, also avois directly setting pos inside `TerrainChanged`, that risks many sync updates per unit.

Also, I have changed the fit function from `GetHeightReal` to `GetApproximateHeight`, this seems to correspond better with how terrain drawing gets updated, also, `GetHeightReal` seems to be updated in an interpolated way, so it needs the unit to "keep listening" while changes are running, with `GetApproximateHeight` it just needs one update per unit.

See above for videos of how both methods look. If terrain rendering was to be updated more smoothly, like `GetHeightReal` is, then the code here would have to be changed too, probably.

So, this introduces some work at update for all units, but with changes from https://github.com/beyond-all-reason/RecoilEngine/pull/2182#issuecomment-2839318561 it can likely be skipped with a global per-unit "moving" flag.

sprunk commented:
> terraforming, since that happens inside SlowUpdate

Would be nice to check what happens if terrain gets smoothly modified by Lua in `gadget:GameFrame`.

saurtron commented:
> Would be nice to check what happens if terrain gets smoothly modified by Lua in `gadget:GameFrame`.

That goes even before that, updated the list in the [above comment](https://github.com/beyond-all-reason/RecoilEngine/pull/2240#issuecomment-2841502665) to reflect that. Should be fine, but I'll setup a test to check the behavior.

sprunk reviewed (commented):
Shaping up good so far.

Perhaps before addressing the comments below it would be good moment to measure the perf benefits (especially in a large real battle with tons of buildings and terrain-deforming explosions, and perhaps bomb/nuke a dense windgen or nano farm for a reproducible artificial-but-not-completely-unrealistic scenario).

saurtron commented on rts/Sim/MoveTypes/StaticMoveType.cpp:
```diff
@@ -29,20 +32,50 @@ CStaticMoveType::CStaticMoveType(CUnit* unit) : AMoveType(unit) {
+	for (auto unit: units) {
+		auto staticMoveType = dynamic_cast<CStaticMoveType*>(unit->moveType);
+		if (staticMoveType) {
+			staticMoveType->needsUpdate = true;
```
Yeah thinking where I might need this (so far I didn't see issues related to that, but I guess maybe lua moving units can place them in the wrong place).

I think it'll work fine when merged with https://github.com/beyond-all-reason/RecoilEngine/pull/2182#issuecomment-283931856, that has a more general "moving" flag we can use for this. So, unsure if doing much work in this PR is needed.

saurtron commented on rts/Sim/Units/Unit.cpp:
```diff
@@ -653,6 +654,10 @@ void CUnit::Update()
 	ASSERT_SYNCED(pos);
 
+
+	auto staticMoveType = dynamic_cast<CStaticMoveType*>(moveType);
+	if unlikely(staticMoveType && staticMoveType->needsUpdate)
```
maybe the compiler will be smart enough about that so don't really need it?

saurtron commented on rts/Sim/MoveTypes/StaticMoveType.cpp:
```diff
@@ -29,20 +32,50 @@ CStaticMoveType::CStaticMoveType(CUnit* unit) : AMoveType(unit) {
+	return false;
+}
+
+void CStaticMoveType::TerrainChanged(int x1, int y1, int x2, int y2)
+{
+	QuadFieldQuery qfQuery;
+	const float3 mins(x1 * SQUARE_SIZE, 0, y1 * SQUARE_SIZE);
+	const float3 maxs(x2 * SQUARE_SIZE, 0, y2 * SQUARE_SIZE);
```
Just mirroring the CFeatureHandler::TerrainChanged here, x-y for 2d is kind of customary too. Anyways if you prefer like this I'll apply it.

sprunk replied:
If `CFeatureHandler::TerrainChanged` also does XY then let's keep it consistent as-is.

saurtron committed:
Reword comment

Co-authored-by: sprunk <spr.ng@o2.pl>
